### ISSUE TITLE
Performance tests need to be more accessible to wider community

### ISSUE THREAD

**Issue by howardjohn** [istio/tools#80]
In the latest networking group meeting, we discussed requiring/highly recommending performance testing performance sensitive changes in the future. The common response was that people don't know about the tests, and if they do, they don't know how to use them. If we want people to be running these tests, we need to make the experience as frictionless as possible, otherwise code will not get tested.

In order to do this, we should make the tests easy to:

* set up
* understand
* interpret

## Set up

Right now, there is a lot of work required to get all of the tests running. I suspect it would take multiple hours the first time. We should have a clear path to take, through documentation and installation scripts.

## Understand

While users don't need to understand the implementation of tests, they do need to understand what they are doing. Especially important is understanding the impact on their cluster, in terms of changes (modifying/removing Istio?) and usage. A common complaint was that people can't run these tests on a small cluster, but there are many tests that can be run on a small cluster - we should explain this.

We also have a bunch of tests piled in various locations of `/perf`. Organizing these would help make the tests more clear. I think a hierarchy like below would be effective:

/perf
* istio-setup - contains all stuff to set up Istio for performance
* load - service graph, maybe other future load generation tools
* stability - all tests that are long running, meant to test stability of a feature
  * http10
  * graceful-shutdown
  * ...
  * `setup_all.sh`
* benchmarking - two pod test, maybe future tools to benchmark Istio
* other - I am not sure where multicluster-vpc and sds-tests fall. Likely in stability, but maybe other. Will need to look more into them
* `README.md` explaining all of this

## Interpret

Once tests are run, there is no/minimal visibility into whether they are succeeding.

For stability tests, we should be able to tell the user "it is working" or "it is broken".

For load tests/benchmarking, we should explain how to interpret the metrics. If we are requiring this for code changes, maybe we want a script to dump important metrics that can be copied into the PR.

### Action Items

* [ ] Improve discoverability by adding references to documentation (partial fixed in https://github.com/istio/tools/pull/73)
* [ ] Improve ease of setup by adding install scripts (partial fixed in https://github.com/istio/tools/pull/73)
* [ ] Document cluster impact of each test - changes to cluster, if any, and resource requirements (ex - each service graph takes ~2 CPU, so they can figure out how many they can run), as well as other requirements (some need GKE)
* [ ] Refactor test folder layout
* [ ] Add/update READMEs for tests to explain what they do, and how to use them
* [ ] Setup all tests to ensure the scripts/docs are up to date and still work
* [ ] Stability tests need a way to report health (partially done in `/metrics`, but needs more work)
* [ ] Document/script how to interpret load testing/benchmarking scripts

**Comment by ardenxu** [comment]
>>>Stability tests need a way to report health (partially done in /metrics, but needs more work)
we are using "perf/load/" tool to test istio performance, and we know we can greb the metrics data for each http endpoint, as we have multiple namespace, so lots of svces that has metrics. But i found out this piece of metrics data is not available on prometheus. 
is there any other good way to format the metrics data in UI? so we can focus on performance part. 
thanks for help in advance.

**Comment by howardjohn** [comment]
What metrics are you looking at that are not in Prometheus? I think all lf the metrics should be in Prometheus so maybe something is wrong there?

**Comment by ardenxu** [comment]
thanks you john for prompt response. For example,  we can get process_virtual_memory_bytes from Prometheus, but can not get service_incoming_requests_total from Prometheus.  what might be possible reason for this? this blocked us for days. 

# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 2.4948736e+07
# HELP service_incoming_requests_total Number of requests sent to this service.
# TYPE service_incoming_requests_total counter
service_incoming_requests_total 1.256637e+06
# HELP service_outgoing_request_size Size in bytes of requests sent from this service.
# TYPE service_outgoing_request_size histogram
service_outgoing_request_size_bucket{destination_service="svc00-0-0",le="1"} 0
service_outgoing_request_size_bucket{destination_service="svc00-0-0",le="10"} 0

**Comment by howardjohn** [comment]
What is the service_incoming_requests_total metric coming from? That isn't one I've seen before , should be looking for istio_requests_total. I am not an expert on all of the metrics though so I could be wrong

**Comment by ardenxu** [comment]
you may mean fortio_requests_total? yes, we can get it from Prometheus.   we can get logs of metrics via curl http://10.104.240.192:8080/metrics query, so by default i thought we can get all of the metrics data from Prometheus UI.

**Comment by howardjohn** [comment]
Something seems wrong here, I think these are metrics from the services, but we should be able to get metrics from the Sidecar proxies (envoy). Do the pods have Sidecars?

**Comment by ardenxu** [comment]
hi john, one more question: is there any way to see performance testing's QPS,  total success rate and process delay with QPS increase?  thank you.

**Comment by ardenxu** [comment]
yes, the svc pods have istio's Sidecars.

**Comment by howardjohn** [comment]
Yes you should he able to get all of these metrics from the Sidecars which will record all sorts of metrics. If you install grafana with Istio it comes with all sorts of dashboards premade with all of these metrics

**Comment by ardenxu** [comment]
hi mandarijog, thanks for echo.  we changed this  Prometheus_scrape=true and re-test the same, but still can not get all service's metrics from prometheus.  prometheus can only report fortio client's related metrics.     we did not config mixed in Istio deployment, should that be a possible reason? thanks.
